Envoy integration

jrajahalme opened the pull request:
This PR adds two Envoy-related subdirectories to Cilium repo:

 envoy: Cilium filters for Envoy. Running tests verifies that Envoy binary was successfully built.

pkg/envoy: Go package for interfacing with Envoy.

Note that for time being the envoy and envoy-api dependencies point to
my forks on Github as we depend on additional functionality not yet
available in Envoy upstream.

To build and use Envoy the environment variable CILIUM_USE_ENVOY needs to be set. CI build adds a parallel step for building cilium with Envoy and running the runtime tests.

```release-note
Technology preview for Envoy as Cilium HTTP proxy, adding HTTP2 and gRPC support.
```

raybejjani commented on envoy/accesslog.cc:
```diff
@@ -0,0 +1,159 @@
+	  kv->set_key(key.c_str());
+	  kv->set_value(value);
+	}
+      }, &entry);
```
Since you only compare the value of key, could this be a switch?

jrajahalme replied:
Not in C or C++ as we are comparing strings rather than integers or enum types.

raybejjani replied:
ah, right, and the pointers aren't guaranteed to be the same. Too many languages :/

raybejjani commented on pkg/envoy/envoy.go:
```diff
@@ -0,0 +1,191 @@
+					}
+				}
+				log.Info("Envoy: Access log closing")
+				conn.Close()
```
I'm pretty sure it won't matter but if you defer this at the start of the function we'll close the connection on a panic/os.Exit. In all those cases, though, the OS will see the close and the unlink on line 114 would happen, so it's probably fine.

jrajahalme replied:
I actually had missed closing the listening socket when stopping envoy. I'll add this.

raybejjani commented on pkg/envoy/envoy.go:
```diff
@@ -0,0 +1,191 @@
+			log.Info("Envoy: Access log connection opened")
+			go func(conn *net.UnixConn) {
+				buf := make([]byte, 4096)
+				for {
```
Since we have a for loop in a goroutine in a for loop it might be clearer to have a label on this for loop that we reference in the break/continue calls below. It's absolutely redundant but gives the reader a stricter reference.
Alternatively, it might be easier to parse these two functions if they were standalone real functions.

jrajahalme replied:
I'll refactor the inner loop as a separate function.

raybejjani commented on pkg/envoy/grpc.go:
```diff
@@ -0,0 +1,629 @@
+
+	ldsServer.glds = grpc.NewServer()
+
+	ldsServer.listenerProto = &envoy_api.Listener{
```
I doubt we can avoid having a complex struct literal like this. As a crazy thought: would it be easier to read if it was json and we parsed it? (we'd have to template the json, so perhaps that's another problem). The positive I see there is that the json we write here looks similar to the one we'd send to envoy and this might make debugging easier.

jrajahalme replied:
I agree that writing grpc that wraps protobuf structs is not easy to read (or write). Maybe a best way forward would be to write the straight protobuf part (that is not generic like filter configuration is) and then use son representation for the filter configuration which would then be translated into the protobuf representation. Envoy does something very similar internally as well.

That said, I'd defer that work to a later PR.

aanm replied:
Don't forget to file a GH issue :-)

raybejjani commented on pkg/envoy/grpc.go:
```diff
@@ -0,0 +1,629 @@
+func (s *LDSServer) updateListener(name string, l7rules policy.L7DataMap) {
+	s.listenersMutex.Lock()
+	log.Debug("Envoy: updateListener ", name)
+	defer s.listenersMutex.Unlock()
```
For a moment I couldn't find the unlock because it wasn't at the end of the function nor deferred right after the lock. I expected the Lock/defer Unlock to be right after each other (despite it making no difference to the code flow in this case). On the other hand, this is the only function that uses defer.

jrajahalme replied:
I'll move the defer one line up. removeLIstener() cannot use defer as the last line (s.bumpVersion()) cannot be called with s.listenersMutex held, due to other call paths locking in the opposite order.

aanm replied:
Isn't the mutex only being used because of line 179? (`l := s.listeners[name]`)

raybejjani commented on pkg/envoy/grpc.go:
```diff
@@ -0,0 +1,629 @@
+		return 0, nil, err
+	}
+	log.Debug("Envoy: RDS Stream DiscoveryRequest ", req.String())
+	version, _ := strconv.ParseUint(req.VersionInfo, 10, 64)
```
Should we handle the error here and return it? We already return errors from this function so I imaging all the callers expect errors.

jrajahalme replied:
Unparsable versionInfo strings will be treated as zeros, and we actually initially depend on this, i.e., an empty string will be treated as number zero. I'll add a comment to clarify this.

raybejjani commented on pkg/envoy/grpc.go:
```diff
@@ -0,0 +1,629 @@
+		return 0, err
+	}
+	log.Debug("Envoy: LDS Stream DiscoveryRequest: ", req.String())
+	version, _ := strconv.ParseUint(req.VersionInfo, 10, 64)
```
Similar question on error handling as the one in RDSServer.recv above

jrajahalme replied:
Similar answer :-)

raybejjani commented on pkg/proxy/envoyproxy.go:
```diff
@@ -0,0 +1,142 @@
+		envoyProxy = envoy.StartEnvoy(true, 0, viper.GetString("state-dir"), viper.GetString("state-dir"))
+	})
+
+	if envoyProxy != nil {
```
Can this race with the envoyOnce above? I can't think of a way to sequence this without a lock, unfortunately :/

jrajahalme replied:
According to sync spec "no call to Do returns until the one call to f returns", so the locking here is implicit.

raybejjani replied:
ooh, I didn't know that! really handy to know.

raybejjani commented on pkg/proxy/logrecord.go:
```diff
@@ -79,3 +78,8 @@ func (l *HTTPLogRecord) log(typ accesslog.FlowType, verdict accesslog.FlowVerdic
+
+func (l *HTTPLogRecord) log(typ accesslog.FlowType, verdict accesslog.FlowVerdict, code int, info string) {
+	l.Timestamp = time.Now().UTC().Format(time.RFC3339Nano)
+	l.log(typ, verdict, code, info)
```
Is this an infinite loop? I think it should be logStamped (above) but maybe there's another layer of indirection I missed.

jrajahalme replied:
Oops, thanks for spotting this. Turns out I only tested with Envoy, hence this was never called. Changed to logStamped().

raybejjani reviewed (changes requested):
Nothing too blocking, except one infinite recursion case. Feel free to override my "Request Changes".
I had some general questions:
- I notice the test program (and the functions it calls) doesn't return any errors. How does it fail? Are we parsing the logs for error messages?
- In a similar vein, when we start envoy, we don't seem to double check that it came up successfully. Since we know it will contact the LDS server after startup it might be useful to wait on that first connection? This would allow us to assert that it's ready to route traffic before we being sending it traffic. Then again, we still have a race between the call to the LDS server and envoy effecting the new listeners (which could be slow on first startup)
- We use bazel's local_repository, pointed at the git submodule's location for envoy and envoy-api. We could use a git_repository reference (https://docs.bazel.build/versions/master/be/workspace.html#git_repository). This behaves the similar to the submodule (fetch at build time) but will be slower if the submodule scheme could reuse the submodule checkout (but that depends on our build system). Once our changes are merged, I would argue that doing all the stuff in Bazel would be better since that's what envoy uses. I'm not terribly attached to this idea, though.

aanm commented on daemon/main.go:
```diff
@@ -292,6 +293,8 @@ func init() {
 	flags.String("enable-policy", endpoint.DefaultEnforcement, "Enable policy enforcement")
 	flags.BoolVar(&enableTracing,
 		"enable-tracing", false, "Enable tracing while determining policy (debugging)")
+	flags.BoolVar(&useEnvoy,
```
We should switch this flag to something similar with what we have in the KVStore (`kvstore` and `kvstore-opt`), this way we can later on add more proxies and proxy options.

jrajahalme replied:
I think the plan is to get rid of this flag before 1.0.

aanm commented on pkg/envoy/envoy.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func createConfig(filePath string, adminAddress string) {
+	config := string("{\n" +
```
tip: In go you can write literal strings with ` in between. (This way you don't need to escape the ")

jrajahalme replied:
Thanks for the hint! Changed :-)

aanm commented on pkg/envoy/grpc.go:
```diff
@@ -0,0 +1,596 @@
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/golang/protobuf/proto"
```
Needs space between line 14 and 15

jrajahalme replied:
OK.

aanm commented on pkg/envoy/grpc.go:
```diff
@@ -0,0 +1,596 @@
+	l := s.listeners[name]
+	// Bail out if this listener does not exist
+	if l == nil {
+		log.Fatalf("Envoy: removeListener: Listener %s does not exist", name)
```
I know it's a fatal but the `s.listenersMutex.Unlock()` is missing here :)

btw, should this be fatal?

jrajahalme replied:
Updating or removing a redirect that does not exist would be an internal logic violation in cilium-agent and should be noticed rather than buried in logs.

If these can fail then they should also return an error, and the callers should be changed to deal with the failures. Currently these are called when bpf is being regenerated (part of the writeHeaderfile()) and we would rather not fail at that point.

aanm commented on pkg/envoy/grpc.go:
```diff
@@ -0,0 +1,596 @@
+	"strconv"
+	"strings"
+
+	envoy_api "github.com/cilium/cilium/pkg/envoy/api"
```
`envoy_api "github.com/cilium/cilium/pkg/envoy/api"` where's this package?

jrajahalme replied:
It is generated from Envoy API .proto -files.

aanm commented on pkg/envoy/grpc.go:
```diff
@@ -0,0 +1,596 @@
+	routes := make([]*envoy_api.Route, 0, len(listener.l7rules))
+	for _, ep := range listener.l7rules {
+		// XXX: We should translate the fromEndpoints selector
+		// (the key of the l7rules map) to a filter in Envoy
```
A filter to receive traffic for the specified endpoint selector?

jrajahalme replied:
Yes, using the numeric source security identity. This is an TODO item for combined L3/4/7 policy.

aanm commented on pkg/proxy/envoyproxy.go:
```diff
@@ -0,0 +1,88 @@
+
+// EnvoyRedirect implements the Redirect interface for a l7 proxy
+type EnvoyRedirect struct {
+	id     string
```
Which id is this? the cilium's endpoint ID?

jrajahalme replied:
It is the redirect's identifier (a parameter given to proxy.CreateOrUpdateRedirect()), which we use when building the identifiers in Envoy configuration so that we can map between Envoy logs and Cilium logs if need be.

aanm commented on pkg/proxy/envoyproxy.go:
```diff
@@ -0,0 +1,153 @@
+	return r.toPort
+}
+
+// ToPort returns the redirect port of an OxyRedirect
```
> // ToPort returns the redirect port of an OxyRedirect

Wrong function and proxy name :-)

aanm commented on pkg/envoy/envoy.go:
```diff
@@ -1,25 +1,36 @@
+	"syscall"
 
 	"github.com/cilium/cilium/pkg/policy"
+	"github.com/golang/protobuf/proto"
```
Space between line 13 and 14

jrajahalme replied:
Added.

aanm commented on pkg/envoy/envoy.go:
```diff
@@ -1,25 +1,36 @@
+}
+
+type EnvoyLogger interface {
+	Log(entry *HttpLogEntry)
```
`HttpLogEntry` doesn't exist

jrajahalme replied:
HttpLogEntry type is generated from accesslog.proto, and will exist after envoy has been built.

aanm commented on pkg/envoy/envoy.go:
```diff
@@ -80,6 +94,74 @@ func StartEnvoy(debug bool, adminPort int, stateDir, logDir string) *Envoy {
+				for {
+					n, _, flags, _, err := conn.ReadMsgUnix(buf, nil)
+					if err != nil {
+						if !isEOF(err) {
```
`err != io.EOF` doesn't work?

jrajahalme replied:
No. The error is more specific with "EOF" at the end of the error string.

aanm commented on pkg/envoy/envoy.go:
```diff
@@ -80,6 +94,74 @@ func StartEnvoy(debug bool, adminPort int, stateDir, logDir string) *Envoy {
+				conn.Close()
+			}(uc)
+		}
+	}(e.accessLogListener)
```
Out of curiosity, why did you pass this listener to the goroutine but not `e.lds`?

jrajahalme replied:
I guess I have mixed styles here... maybe I thought the "e.accessLogListener" is long or something :-). I'll remove the explicit parameter passing.

aanm commented on pkg/envoy/grpc.go:
```diff
@@ -289,29 +297,46 @@ func (s *RDSServer) translatePolicyRule(h api.PortRuleHTTP) *envoy_api.Route {
 	headers := make([]*envoy_api.HeaderMatcher, 0, cnt)
 	if h.Path != "" {
 		headers = append(headers, &envoy_api.HeaderMatcher{Name: ":path", Value: h.Path, Regex: &isRegex})
+		rule_ref = "PathRegexp(\"" + h.Path + "\")"
```
String literal tip I gave back can be applied here.

jrajahalme replied:
Yes, thanks!

aanm commented on pkg/proxy/envoyproxy.go:
```diff
@@ -86,3 +95,59 @@ func (r *EnvoyRedirect) Close() {
+		}
+	}
+
+	url := url.URL{
```
Try avoiding variable names that have package names

jrajahalme replied:
fixed. Also fixed the Oxy -references in comments you mentioned above.

aanm commented on Vagrantfile:
```diff
@@ -116,8 +116,12 @@ Vagrant.configure(2) do |config|
+	    #  -c: Use checksum to prevent copying of files (avoid unnecessary bazel rebuilds)
             config.vm.synced_folder '.', '/home/vagrant/go/src/github.com/cilium/cilium', type: "rsync",
-            rsync__exclude: ["contrib/packaging/docker/stage", "src"]
+            rsync__exclude: ["contrib/packaging/docker/stage", "src"], rsync__args: "-lc"
```
@ianvernon @eloycoto we probably need this for the test images?

jrajahalme replied:
I'm trying out having bazel manage the Envoy dependency. This will remove the need for the linked files, so this should not be needed going forward.

aanm reviewed (changes requested):
b4776c55d8f020733995004be53b9efd10409dfb doesn't seem to have anything

Should we move envoy/ and submodules to pkg/envoy?

jrajahalme committed:
CI: Runtime tests with Envoy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
tests: Prepare for Envoy integration.

Ignore Envoy dependencies and intermediate build files in format tests.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Envoy integration.

This commit adds two Envoy-related subdirectories to Cilium repo:

envoy: Cilium filters, built with bazel that pulls in Envoy. Running tests
       verifies that Envoy binary was successfully built.

pkg/envoy: Go package for interfacing with Envoy.

pkg/proxy/envoyproxy.go contains the Cilium redirect code that uses
Envoy.  Cilium daemon needs to be started with command line option
"--envoy-proxy" to use Envoy instead of Oxy for HTTP proxy.

Note that for time being the envoy and envoy-api dependencies point to
my forks on Github as we depend on additional functionality not yet
available in Envoy upstream.

TODO:

1. Envoy log rotation (not sure if Envoy already does this)

2. Notification of deployed envoy configuration back to Cilium
(CLI). Envoy configuration updates are asynchronous, and while the
cilium envoy package gets both positive and negative acknowledgements
back from Envoy, that information is currently not propagated
anywhere.

3. Pooling of Envoy configuration updates. Envoy listener
configuration updates all listeners at once, while Cilium calculates
those updates one-by-one. Currently this leads to N Envoy Listener
updates for a single policy update involving N redirects.

4. Do not create separate redirects for same set of rules on multiple
(TCP) ports. Currently Cilium creates a new redirect for each allowed
port, even if the set of rules is the same. Since the original
destionation port is preserved via the bpf proxy map, there is no need
to create separate proxy redirects/listeners for different ports on
the same endpoint that share the L7 configuration.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off-by: Thomas Graf <thomas@cilium.io>

jrajahalme committed:
vendor: Add and update go dependencies.

envoy gRPC API has dependency on proto-gen-go, various ptypes,
grpc/reflection and genproto, which in turn needs an update on
protobuf dependency.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Envoy: Set original destination cluster cleanup interval to 1.5 seconds.

Envoy keeps upstream connections open for later reuse, which helps
avoid unnecessary TCP handshake roundtrips.  Unfortunately this breaks
down when a policy change changes the real destination of those
connections, e.g., redirecting the connection to a proxy.  In this
case the connection does not work any more, as the new destination
does not have the right connection state.

Orignal destination clusters can open unlimited number of upstream
connections, as the set of upstream hosts is not limited by
configuration.  In order to limit resource use, original destination
clusters periodically scan all the previously used destination hosts
and removes the ones that have not been used for upstream traffic
since the last period.  This period between the cleanups is
configurable and defaults to 5 seconds. This is too long for some of
the Cilium runtime tests, that repetitively issue traffic between
endpoints, but change the network policy in between.

Changing the cleanup interval to 1.5 seconds seems to allow these
tests to pass correctly, while still allowing re-use of the upstream
connections in cases where the gap between requests is no longer than
1.5 seconds.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Add generated protobuf files to the repo

Having the generated files in repo allows for skipping Envoy build if
bazel is missing.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
helpers: Copy logs from /var/run/cilium after test failures.

Copy log files, if any, from /var/run/cilium when gathering files after a failed test.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Use local copies of required BPF header files

Depending on the version of the system headers available when compilation
happens, the version of <linux/bpf.h> available may be insufficient.

There are two possible options to address this:
 1. Refer to the top cilium directory as an external bazel dependency.
 2. Provide a local copy of the header files

Given that these headers are rarely ever going to change, making a local copy
of these header file is chosen for simplicity.

Signed-off-by: Thomas Graf <thomas@cilium.io>

jrajahalme committed:
daemon: Add support for CILIUM_USE_ENVOY environment variable.

jrajahalme committed:
envoy: Make Envoy build conditional on CILIUM_USE_ENVOY.

Build and install Envoy only if the environment variable
CILIUM_USE_ENVOY is set.

Skip the Envoy package unit tests if CILIUM_USE_ENVOY is not set.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
